Halo add corners

TomMelt opened the pull request:
# Add corner neighbours into halo exchange logic

### Task List
- [x] Linked an issue above that captures the requirements of this PR
- [x] Defined the tests that specify a complete and functioning change
- [x] Implemented the source code change that satisfies the tests
- [x] Commented all code so that it can be understood without additional context
- [x] No new warnings are generated or they are mentioned below
- [x] The documentation has been updated (or an issue has been created to do so)
- [x] Relevant labels (e.g., enhancement, bug) have been applied to this PR
- [x] This change conforms to the conventions described in the README

---

# Change Description

This PR has the following key changes:

- adds support for corner neighbours to the existing halo exchange logic (which previously just exchanged edges)
- add support for `CGVector` and `DGVectorHolder` exchange (this is required for corners and edges)
- updates metadata tests and closed-boundary (CB) halo test to work with corner neighbours
- removes the Periodic-Boundary (PB) halo exchange test for now, this will be addressed in a subsequent PR #1044

---
# Test Description

### `HaloExchangePB_test.cpp`

`HaloExchangeCB_test.cpp` creates test data for all the array types:
- `HField`
- `VertexField`
- `DGField`
- `DGVector`
- `CGVector`

I modified the previous version, so that the model data are created in such a way that the value of each cell can be calculated based on its indices. Therefore we can check after exchange that all of the cells have been exchanged successfully. This also now allows us to change the number of ranks arbitrarily (assuming you also provide the appropriate `partition_metadata` file. 

### `ModelMetadataCB_test.cpp` and `ModelMetadataPB_test.cpp`

Small modifications have been made to the periodic and closed boundary `ModelMetadata` tests.

The changes now add the additional corner neighbour metadata and check that it is ready correctly.

### `partition_metadata_3_cb.cdl` and `partition_metadata_3_pb.cdl`

These files are now generated at compile time by running the `decomp` tool on a new file `halo_test_mask.cdl`. This is in line with similar changes introduced by @joewallwork on the XIOS tests.

---
# Other Details

* `domain_decop` git commit has now been bumped to the latest version on main. This version of the decomp tool contains the corner neighbour metadata.

### Further work

- Integrating halo exchange to nextsim is handled by PR #924

TomMelt committed:
feat: add corner neighbours to nextsim

TomMelt committed:
chore: bump commit of domain_decomp

TomMelt committed:
tests: remove Halo PB test for now

TomMelt committed:
chore: remove unused includes

TomMelt committed:
docs: update doxygen comments in Halo.hpp

joewallwork commented on core/src/include/Halo.hpp:
```diff
@@ -52,11 +54,36 @@ class Halo {
 
     /*!
      * @brief Constructs a halo object from DGVector
```
```suggestion
     * @brief Constructs a halo object from DGVectorHolder
```

TomMelt replied:
Thanks done in f1829af1

joewallwork commented on core/src/include/Halo.hpp:
```diff
@@ -52,11 +54,36 @@ class Halo {
+        CGdegree = N;
+        nCells = CGdegree;
         setSpatialDims();
         intializeHaloMetadata();
```
In none of these cases do the constructors depend on their argument, right? The only information that's being used is the value of `N` and the type. So do we actually need to provide `dgv`, `dgvh`, or `cgv` as arguments?

TomMelt replied:
I think we do, because we need to templated `N` from  `dgv` and `dgvh`.

You are right that `Halo(DGVectorHolder<N>& dgvh)` and `Halo(DGVector<N>& dgv)` are essentially the same. But `Halo(ModelArray& ma)` and `Halo(CGVector<N>& cgv)` set different member variables from `dgv`.

What I could do however, is change `Halo(DGVectorHolder<N>& dgvh)` to call `Halo(DGVector<N>& dgv)` with `static_cast<DGVector<DGadvection>&>()` like we do in dynamics:

https://github.com/nextsimhub/nextsimdg/blob/87dadeb5c30e5156015e06761ce3635961f995c9/dynamics/src/include/BrittleCGDynamicsKernel.hpp#L131-L134

TomMelt replied:
I have simplified this in f1829af1

joewallwork commented on core/src/include/Halo.hpp:
There seems to be quite a lot of implementation in this header file. Might it be better to separate it out into `core/src/Halo.cpp`?

Thanduriel replied:
Agreed. Function definitions should go in a source file if possible. Exceptions are (public) template functions which have to be defined in the header and maybe one-liners.

TomMelt replied:
Thanks @joewallwork and @Thanduriel . I have split the implementation as suggested and I also tidied up a bit.

- Split implementation out from `Halo.hpp` into `core/src/Halo.cpp`
- Moved implementation of non-templated functions to `core/src/Halo.cpp`
- Made the following functions private:
  - `setSpatialDims()`
  - `intializeHaloMetadata()`
  - `sendBufferPositions()`
  - `sourceToSendBuffer()`
  - `recvBufferPositions()`
  - `recvBufferToTarget()`
  - `recvBufferToTarget()`
  - `transposeCorners()`
  - `populateTarget()`

See commit c2136348

joewallwork commented on core/src/include/Halo.hpp:
```diff
@@ -234,35 +258,164 @@ class Halo {
-
-        for (auto edge : edges) {
-            size_t beg = std::accumulate(m_edgeLengths.begin(), m_edgeLengths.begin() + edge, 0);
-            size_t num = m_edgeLengths.at(edge);
-
-            Slice sourceSlice = m_innerSlices.at(edge);
-            SliceIter sourceIter(sourceSlice, { m_Nx, m_Ny });
-
-            if (isVertical(edge)) {
-                Slice tempBufferSlice = { { {}, { beg, beg + num } } };
-                // spatial dims are flattened into 1-D.
-                SliceIter tempBufferIter(tempBufferSlice, { 1, m_numHaloCells });
-                ModelArraySlice::copySliceWithIters(source, sourceIter, tempBuffer, tempBufferIter);
-            } else {
-                Slice tempBufferSlice = { { { beg, beg + num }, {} } };
-                // spatial dims are flattened into 1-D.
-                SliceIter tempBufferIter(tempBufferSlice, { m_numHaloCells, 1 });
-                ModelArraySlice::copySliceWithIters(source, sourceIter, tempBuffer, tempBufferIter);
+        size_t buffer_len = sendBufferSize;
+        if (isCG) {
+            buffer_len = sendBufferSize / CGdegree;
+        }
+        for (size_t comp = 0; comp < m_numComps; ++comp) {
+            Eigen::Map<Eigen::ArrayXXd, 0, Eigen::InnerStride<Eigen::Dynamic>> source_map(
+                source.col(comp).data(), m_Nx, m_Ny, Eigen::InnerStride<>(m_numComps));
+            Eigen::Map<Eigen::ArrayXXd> buffer_map(send[comp].data(), buffer_len, nCells);
+            for (auto edge : edges) {
```
This code could benefit from some comments as I'm struggling to understand what it does. numComps is the `N` value in the template constructor, right?

TomMelt replied:
Good point. I have added explanation here 0436038d

Please let me know if this is clearer.

joewallwork replied:
This is much clearer, thanks!

joewallwork commented on core/src/include/Halo.hpp:
```diff
@@ -234,35 +258,164 @@ class Halo {
+     * @brief Calculate which edge the send position sendPos would fall under in the send buffer
+     *
+     */
+    Edge edgeFromSendPos(int sendPos, int fromRank)
```
Missing `@params` etc in docstring

TomMelt replied:
Thanks see eb26a814

joewallwork commented on core/src/include/Halo.hpp:
```diff
@@ -234,35 +258,164 @@ class Halo {
+            count = 1;
+            disp = disp + sendEdge;
+            recvOffset = recvOffset + 4;
+            // this is messy
```
Is this a TODO to do it differently?

TomMelt replied:
Nope, this is the best way I could think to do it. But I have updated the comment to be more helpful

a8c47722

joewallwork commented on core/src/include/Halo.hpp:
```diff
@@ -374,9 +516,235 @@ class Halo {
+                // periodic case
+                hasCorner = true;
+                isPeriodic = true;
+            }
```
Presumably it's impossible for both conditions to be true at the same time? Or would it be useful to include a failure mode in case it happens due to a bug?

TomMelt replied:
Sure, that's a good point. It shouldn't be possible. So I will add a check in case.

see c030bf38

joewallwork commented on core/src/include/Halo.hpp:
```diff
@@ -374,9 +516,235 @@ class Halo {
+        }
+    }
+
+    template <typename T> void populateTarget(T& target)
```
Please add a docstring.

TomMelt replied:
Thanks for spotting. Done in c76f111e

joewallwork commented on core/src/ModelMetadata.cpp:
```diff
@@ -148,6 +195,24 @@ void ModelMetadata::getPartitionMetadata(std::string partitionFile)
 
     readNeighbourData(ncFile);
 
+    // cornerHaloRecv doesn't need to be read because it can be easily calculated.
```
I guess there's an assumption (probably right) that a small amount of computation is cheaper than a small amount of I/O?

TomMelt replied:
That's true, but the main reason is it would require a lot of changes in `domain_decomp` to essentially add a fixed constant which is essentially 1 (i.e., the `halo_size`). We would need to write routines to save that metadata in `domain_decomp` and read it in `nextsim`. I thought it wasn't worth the effort :sweat_smile:

joewallwork commented on core/test/HaloExchangeCB_test.cpp:
Nice to reduce the amount of numbers hard-coded in code here, well done!

TomMelt replied:
Thank you :)

joewallwork reviewed (commented):
Thanks for this @TomMelt. Here's some initial feedback. I haven't checked all of the logic but let me know if there's anything in particular you want checking.

Thanduriel commented on core/src/include/ModelMetadata.hpp:
```diff
@@ -206,6 +222,14 @@ class ModelMetadata {
     neighbourArray neighbourHaloSendPeriodic;
     neighbourArray neighbourHaloRecvPeriodic;
+
+    typedef std::array<std::vector<int>, N_CORNER> cornerArray;
```
Types should be in CamelCase.
```
using CornerArray = std::array<std::vector<int>, N_CORNER>;
```

Thanduriel commented on core/src/include/ModelMetadata.hpp:
```diff
@@ -191,12 +191,28 @@ class ModelMetadata {
+    std::array<std::string, N_CORNER> cornerNames
+        = { "top_left", "top_right", "bottom_right", "bottom_left" };
+
     typedef std::array<std::vector<int>, N_EDGE> neighbourArray;
```
Types should be in CamelCase. Prefer `using` over `typedef` (see [CppCoreGuidlines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#t43-prefer-using-over-typedef-for-defining-aliases)).
```
using NeighbourArray = std::array<std::vector<int>, N_EDGE>;
```

TomMelt replied:
Thanks this and CornerArray have been done in a59341f4

Thanduriel commented on core/src/include/ModelMetadata.hpp:
```diff
@@ -191,12 +191,28 @@ class ModelMetadata {
+     * @brief Gets the extents of the grid in the X direction for all ranks.
+     * @return A vector containing the number of grid points in X for each rank.
+     */
+    std::vector<int> getRankExtentsX() const;
```
Return by `const&` to prevent unnecessary copies.

Thanduriel commented on core/src/include/ModelMetadata.hpp:
```diff
@@ -191,12 +191,28 @@ class ModelMetadata {
+     * @brief Gets the extents of the grid in the Y direction for all ranks.
+     * @return A vector containing the number of grid points in Y for each rank.
+     */
+    std::vector<int> getRankExtentsY() const;
```
Return by `const&` to prevent unnecessary copies.

TomMelt replied:
Thanks `getRankExtentsX()` and `getRankExtentsY()` have been modified in c4210bbc

Thanduriel commented on core/src/include/Halo.hpp:
```diff
@@ -374,9 +516,235 @@ class Halo {
+    {
+        auto& metadata = ModelMetadata::getInstance();
+        int fromRank;
+        size_t disp;
```
These variables could be defined closer to where they are used, i.e. in the `if (hasCorner)` branch.

TomMelt replied:
Thanks see bcfedc7d

Thanduriel commented on core/src/include/Halo.hpp:
```diff
@@ -374,9 +516,235 @@ class Halo {
+                    auto buffer_len = recvBufferSize / CGdegree;
+                    for (size_t comp = 0; comp < m_numComps; ++comp) {
+                        Eigen::Map<Eigen::ArrayXXd> rmap(recv[comp].data(), buffer_len, nCells);
+                        auto offset = buffer_len - (4 - corner) * nCells;
```
Where does this 4 come from? Is it just `Corner::N_CORNER`?

TomMelt replied:
Yes, I have replaced this and a few other magic numbers with the enums b6938692

Thanduriel commented on core/src/ModelMetadata.cpp:
```diff
@@ -157,6 +222,8 @@ int ModelMetadata::getLocalExtentX() const { return localExtentX; }
 int ModelMetadata::getGlobalExtentX() const { return globalExtentX; }
 int ModelMetadata::getGlobalExtentY() const { return globalExtentY; }
+std::vector<int> ModelMetadata::getRankExtentsX() const { return rankExtentsX; }
+std::vector<int> ModelMetadata::getRankExtentsY() const { return rankExtentsY; }
```
```suggestion
const std::vector<int>& ModelMetadata::getRankExtentsX() const { return rankExtentsX; }
const std::vector<int>& ModelMetadata::getRankExtentsY() const { return rankExtentsY; }
```

TomMelt replied:
Thanks this was fixed in c4210bbc

Thanduriel commented on core/test/ModelMetadataPB_test.cpp:
```diff
@@ -23,13 +23,19 @@ constexpr ModelMetadata::Edge RIGHT = ModelMetadata::Edge::RIGHT;
+constexpr ModelMetadata::Corner TOP_LEFT = ModelMetadata::Corner::TOP_LEFT;
+constexpr ModelMetadata::Corner TOP_RIGHT = ModelMetadata::Corner::TOP_RIGHT;
+constexpr ModelMetadata::Corner BOTTOM_RIGHT = ModelMetadata::Corner::BOTTOM_RIGHT;
+constexpr ModelMetadata::Corner BOTTOM_LEFT = ModelMetadata::Corner::BOTTOM_LEFT;
```
If the intend here is just to introduce shorter names, `using` would be more clear:
```
using enum ModelMetadata::Corner;
```

TomMelt replied:
Thanks that was my intent. It appears that your suggested syntax is only available in `-std=c++20`, see [here](https://en.cppreference.com/w/cpp/language/enum.html).

Are you happy for me to keep it? or do you have any other suggestions?

Thanduriel replied:
I guess the pre C++20 alternative would be to introduce each name separately which is still a bit more compact than the current version:
```
using ModelMetadata::Corner::TOP_LEFT;
```
But I am fine with the current version.

TomMelt replied:
It looks like `using ModelMetadata::Corner::TOP_LEFT;` is also only available in `-std=c++20` but I have done the following which is slightly better:

```suggestion
using Corner = ModelMetadata::Corner;

constexpr auto TOP_LEFT = Corner::TOP_LEFT;
constexpr auto TOP_RIGHT = Corner::TOP_RIGHT;
constexpr auto BOTTOM_RIGHT = Corner::BOTTOM_RIGHT;
constexpr auto BOTTOM_LEFT = Corner::BOTTOM_LEFT;
```

Thanduriel commented on core/src/include/Halo.hpp:
```diff
@@ -234,35 +258,164 @@ class Halo {
+
+        // extents of sending domain
+        auto extentX = metadata.getRankExtentsX()[fromRank];
+        auto extentY = metadata.getRankExtentsY()[fromRank];
```
I think it would be better to not use `auto` if the expression has a simple type. Here the type is just `int` while in many other places the standard size type `size_t` is used instead. The distinction between signed and unsigned ints is important because the following conditions would not work with an unsigned type.

TomMelt replied:
Thanks that's a very good point. And a footgun I shot myself with several times. Changed to `int` here  23e3ca9c

Thanduriel commented on core/src/include/Halo.hpp:
```diff
@@ -234,35 +258,164 @@ class Halo {
+            auto extentX = CGdegree * metadata.getRankExtentsX()[fromRank] + 1;
+            auto extentY = CGdegree * metadata.getRankExtentsY()[fromRank] + 1;
+            // TODO this is too big. Should only be 2 x CGdegree * haloWidth (extentX + extentY)
+            auto fromRankSendBufferSize = 2 * haloWidth * CGdegree * (extentX + extentY);
```
Is this todo outdated? The line below looks just like suggested.

TomMelt replied:
Yep thanks for spotting it. It was a reminder that is no longer valid. Removed here eeeecf07

Thanduriel reviewed (changes requested):
Here is my feedback.

TomMelt committed:
chore: simplify Halo const for DGVectorHolder

TomMelt committed:
chore: update comment in recvPositions

TomMelt committed:
docs: explain use of Eigen::Map in populateSendBuffers

TomMelt committed:
[skip ci] chore: fix docstring for edgeFromSendPos in Halo.hpp

TomMelt committed:
[skip ci] feat: catch potential error if corner is periodic and non-periodic

TomMelt committed:
[skip ci] chore: add docstring for populateTarget

TomMelt committed:
[skip ci] chore: use using over typedef

TomMelt committed:
[skip ci] chore: return const& instead to prevent copies

TomMelt committed:
[skip ci] chore: move fromRank & disp definition close to usage

TomMelt committed:
[skip ci] chore: replace magic number 4 with N_CORNER/N_EDGE

TomMelt committed:
[skip ci] chore: replace auto with int for extents

TomMelt committed:
[skip ci] chore: remove old comment

TomMelt committed:
chore: split Halo class into header and implementation files

- Split implementation out from `Halo.hpp` into `core/src/Halo.cpp`
- Moved implementation of non-templated functions to `core/src/Halo.cpp`
- Added `Halo.cpp` to `core/src/CMakeLists.txt`
- Made the following functions private:
  - `setSpatialDims()`
  - `intializeHaloMetadata()`
  - `sendBufferPositions()`
  - `sourceToSendBuffer()`
  - `recvBufferPositions()`
  - `recvBufferToTarget()`
  - `recvBufferToTarget()`
  - `transposeCorners()`
  - `populateTarget()`

All tests pass: `ctest` (6/6 tests passed)

Thanduriel reviewed (approved):
I have one alternative suggestion for the shorter enum types but you decide if you want to make the change. From my side the code is ready for merge.

TomMelt committed:
chore: simplify enums in ModelMetadata tests

joewallwork commented on core/src/include/Halo.hpp:
```diff
@@ -402,47 +465,14 @@ class Halo {
-     * │ │x│x│ │ (empty) = unused data in DGVector
-     * └─┴─┴─┴─┘
-     */
     template <typename T> void exchangeHalos(T& target)
```
Please could you add a docstring for this function?

TomMelt replied:
Thanks changed in 876dea49

joewallwork commented on core/src/Halo.cpp:
```diff
@@ -0,0 +1,390 @@
+ *
+ * Halo supports the main data structures of NextSim e.g., ModelArray, DGVector and CGVector.
+ *
+ * The halos are exchange via one-sided MPI communication using RMA.
```
Perhaps spell out RMA here for clarity.

TomMelt replied:
Thanks changed in 876dea49

joewallwork commented on core/src/Halo.cpp:
```diff
@@ -0,0 +1,390 @@
+
+void Halo::populateRecvBuffers()
+{
+
```
```suggestion
```

TomMelt replied:
Thanks changed in 876dea49

joewallwork reviewed (approved):
This is great, thanks @TomMelt. I have a few small requests but I think it's otherwise good to go!

TomMelt committed:
chore: small changes suggested by JW
